diff --git a/node_modules/lunr-languages/lunr.zh.js b/node_modules/lunr-languages/lunr.zh.js
index 48f5890..d377a2c 100644
--- a/node_modules/lunr-languages/lunr.zh.js
+++ b/node_modules/lunr-languages/lunr.zh.js
@@ -30,18 +30,18 @@
      * only CommonJS-like environments that support module.exports,
      * like Node.
      */
-    module.exports = factory(require('@node-rs/jieba'))
+    module.exports = factory(null) // nodejieba is not available in RN, we'll use a polyfill and pass it through the constructor function
   } else {
     // Browser globals (root is window)
     factory()(root.lunr);
   }
-}(this, function(nodejieba) {
+}(this, function(_nodejieba /* This will be null */) {
   /**
    * Just return a value to define the module export.
    * This example returns an object, but the module
    * can return a function as the exported value.
    */
-  return function(lunr, nodejiebaDictJson) {
+  return function(lunr, nodejiebaDictJson, nodejieba /* pass the nodejieba polyfill here */) {
     /* throw error if lunr is not yet included */
     if ('undefined' === typeof lunr) {
       throw new Error('Lunr is not present. Please include / require Lunr before this script.');
@@ -121,6 +121,9 @@
         }
       });
     }
+    // Need to set this otherwise lunr.zh.tokenizer will not be used at all. (NOTE: This will affect all lunr tokenizations!)
+    lunr.tokenizer = lunr.zh.tokenizer;
+    this.tokenizerFn = lunr.zh.tokenizer;
 
     /* lunr trimmer function */
     lunr.zh.wordCharacters = "\\w\u4e00-\u9fa5";
@@ -139,7 +142,7 @@
 
     /* lunr stop word filter. see https://www.ranks.nl/stopwords/chinese-stopwords */
     lunr.zh.stopWordFilter = lunr.generateStopWordFilter(
-      '的 一 不 在 人 有 是 为 為 以 于 於 上 他 而 后 後 之 来 來 及 了 因 下 可 到 由 这 這 与 與 也 此 但 并 並 个 個 其 已 无 無 小 我 们 們 起 最 再 今 去 好 只 又 或 很 亦 某 把 那 你 乃 它 吧 被 比 别 趁 当 當 从 從 得 打 凡 儿 兒 尔 爾 该 該 各 给 給 跟 和 何 还 還 即 几 幾 既 看 据 據 距 靠 啦 另 么 麽 每 嘛 拿 哪 您 凭 憑 且 却 卻 让 讓 仍 啥 如 若 使 谁 誰 虽 雖 随 隨 同 所 她 哇 嗡 往 些 向 沿 哟 喲 用 咱 则 則 怎 曾 至 致 着 著 诸 諸 自'.split(' '));
+      '的 一 不 在 人 有 是 為 以 於 上 他 而 後 之 來 及 了 因 下 可 到 由 這 與 也 此 但 並 個 其 已 無 小 我 們 起 最 再 今 去 好 只 又 或 很 亦 某 把 那 你 乃 它 吧 被 比 別 趁 當 從 到 得 打 凡 兒 爾 該 各 給 跟 和 何 還 即 幾 既 看 據 距 靠 啦 了 另 麼 每 們 嘛 拿 哪 那 您 憑 且 卻 讓 仍 啥 如 若 使 誰 雖 隨 同 所 她 哇 嗡 往 哪 些 向 沿 喲 用 於 咱 則 怎 曾 至 致 著 諸 自 的 一 不 在 人 有 是 爲 以 於 上 他 而 後 之 來 及 了 因 下 可 到 由 這 與 也 此 但 並 個 其 已 無 小 我 們 起 最 再 今 去 好 只 又 或 很 亦 某 把 那 你 乃 它 吧 被 比 別 趁 當 從 到 得 打 凡 兒 爾 該 各 給 跟 和 何 還 即 幾 既 看 據 距 靠 啦 了 另 麼 每 們 嘛 拿 哪 那 您 憑 且 卻 讓 仍 啥 如 若 使 誰 雖 隨 同 所 她 哇 嗡 往 哪 些 向 沿 喲 用 於 咱 則 怎 曾 至 致 着 諸 自 的 一 不 在 人 有 是 为 以 于 上 他 而 后 之 来 及 了 因 下 可 到 由 这 与 也 此 但 并 个 其 已 无 小 我 们 起 最 再 今 去 好 只 又 或 很 亦 某 把 那 你 乃 它 吧 被 比 别 趁 当 从 到 得 打 凡 儿 尔 该 各 给 跟 和 何 还 即 几 既 看 据 距 靠 啦 了 另 么 每 们 嘛 拿 哪 那 您 凭 且 却 让 仍 啥 如 若 使 谁 虽 随 同 所 她 哇 嗡 往 哪 些 向 沿 哟 用 于 咱 则 怎 曾 至 致 着 诸 自'.split(' '));
     lunr.Pipeline.registerFunction(lunr.zh.stopWordFilter, 'stopWordFilter-zh');
   };
 }))
