--- a/cert_checker.rb
+++ b/cert_checker.rb
@@ -116,7 +116,7 @@ module FastlaneCore

       # Find all installed WWDRCA certificates
       installed_certs = []
-      Helper.backticks("security find-certificate -a -c '#{certificate_name}' -p #{wwdr_keychain.shellescape}")
+      Helper.backticks("security find-certificate -a -c '#{certificate_name}' -p #{wwdr_keychain.shellescape}", print: false)
             .lines
             .each do |line|
         if line.start_with?('-----BEGIN CERTIFICATE-----')
